src/txHashBuilder.c

```diff
@@ -288,20 +292,347 @@ void txHashBuilder_addCertificate_delegation(
+	}
+
+	builder->state = TX_HASH_BUILDER_IN_CERTIFICATES_POOL_OWNERS;
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_addOwner(
+        tx_hash_builder_t* builder,
+        const uint8_t* stakingKeyHash, size_t stakingKeyHashSize
+)
+{
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_addOwner: %d", builder->state);
+
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_OWNERS);
+	ASSERT(builder->poolCertificateData.remainingOwners > 0);
+	builder->poolCertificateData.remainingOwners--;
+
+	// Bytes[poolKeyHash]
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, stakingKeyHashSize);
+		BUILDER_APPEND_DATA(stakingKeyHash, stakingKeyHashSize);
+	}
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_enterRelays(tx_hash_builder_t* builder)
+{
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_enterRelays: %d", builder->state);
+
+	// enter empty owners if none were received (and none were expected)
+	if (builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_PARAMS) {
+		ASSERT(builder->poolCertificateData.remainingOwners == 0);
+		txHashBuilder_addPoolRegistrationCertificate_enterOwners(builder);
+	}
+
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_OWNERS);
+	ASSERT(builder->poolCertificateData.remainingOwners == 0);
+
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, builder->poolCertificateData.remainingRelays);
+	}
+
+	builder->state = TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS;
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_addRelay0(
+        tx_hash_builder_t* builder,
+        const uint16_t* port,
+        const ipv4_t* ipv4,
+        const ipv6_t* ipv6
+)
 {
-	ASSERT(builder->state <= TX_HASH_BUILDER_IN_CERTIFICATES);
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_addRelay0: %d", builder->state);
 
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS);
+	ASSERT(builder->poolCertificateData.remainingRelays > 0);
+	builder->poolCertificateData.remainingRelays--;
+
+	// Array(4)[
+	//   Unsigned[0]
+	//   Unsigned[port] / Null
+	//   Bytes[ipv4] / Null
+	//   Bytes[ipv6] / Null
+	// ]
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 4);
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, 0);
+		}
+		{
+			if (port != NULL) {
+				BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, *port);
+			} else {
+				BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
+			}
+		}
+		{
+			if (ipv4 != NULL) {
+				ASSERT(sizeof(ipv4->ip) == IPV4_SIZE); // SIZEOF does not work for 4-byte buffers
+				BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, IPV4_SIZE);
+				BUILDER_APPEND_DATA(ipv4->ip, IPV4_SIZE);
+			} else {
+				BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
+			}
+		}
+		{
+			if (ipv6 != NULL) {
+				ASSERT(sizeof(ipv6->ip) == IPV6_SIZE); // SIZEOF does not work for 4-byte buffers
+				BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, IPV6_SIZE);
+
+				// serialized as 4 big-endian uint32
+				for (size_t i = 0; i < 4; i++) {
+					uint8_t chunk[4];
+					uint32_t* valuePtr = (uint32_t *)(ipv6->ip + 4 * i);
```

Comment:
could be `uint32_t* as_uint32 = ipv6->ip` and then index directly into `as_uint32`